onnxscript/rewriter/pattern.py

```diff
@@ -985,6 +985,13 @@ def __init__(self, rules: Sequence[RewriteRule], *, commute: bool = False) -> No
         if commute:
             rules = list(itertools.chain.from_iterable([rule.commute() for rule in rules]))
         self.rules = rules
+        self.rule_mapping = {}
+        for rule in self.rules:
+            target_node_op_type = rule._target_node_pattern.op._value
```

Comment:
We should not assume that the target node pattern will have a definitive op value. (A recent PR allows us to create a node pattern that will match against all calls to ops/functions with a name that matches a prefix for example.)

However, in the common case it will. Hence, to optimize this, we will need a combined scheme: for the common case we should partition the rules using a dictionary (as in this PR). But we should have a fallback list of rules to be applied to all nodes (when the rule doesn't correspond to a single optype).